### ISSUE TITLE
[BUG][mthreads]Valid 2D max-pooling kernel with semantics-no-op dynamic branches is miscompiled

### ISSUE THREAD

**Issue by O1iveira7** [FlagTree/flagtree#1009]
##  Summary 
A semantically equivalent variant of a valid Triton kernel (42_2d_max_pooling) is silently miscompiled by the MTGPU backend: the kernel compiles and runs, but the output of the untouched main path is wrong at 24/25 positions (only index 0 coincides, by chance). The injected code is provably semantics no-op — it only writes to a dedicated scratch ("switch")buffer and both branch predicates are compile-time-uniform — yet the main-path masked load + reduce is corrupted. The output value set collapses to {window-0 max replicated, -inf}, where -inf is the other= padding value of the masked load leaking through — i.e. the load masks themselves are being computed incorrectly for some lanes.
## Environment
   • Triton 3.6.0 (FlagTree mthreads3.6, x64)
   • torch 2.7.1 / torch_musa 2.7.1 / MUSA driver 40305
   • Hardware: MTT S5000

**FlagTree mthreads3.6 output:**
>original   = [2.265505075454712, ..., 0.955923318862915]
>transformed= [2.265505075454712, ..., -inf]
>mismatch = 21/25

**Triton3.7 RTX3070 output:**
>original    = [2.265504837036133, ...,0.9559233784675598]
>transformed = [2.265504837036133,..., 0.9559233784675598]
>mismatch = 0/25

  ## Minimal reproducer
   ```python
     import torch
     import triton
     import triton.language as tl

     # functional[3] params of 42_2d_max_pooling
     H, W, KS, STRIDE, PADDING = 10, 10, 5, 2, 2
     BLOCK_SIZE_KERNEL = triton.next_power_of_2(KS)  # 8
     BLOCK_SIZE_OUTPUT = 1024
     NUM_OUT = 25  # (10+4-5)//2+1 == 5 per dim


     # ============================ original kernel (correct) ============================
     @triton.jit
     def max_pooling_kernel(
         input: tl.tensor, output: tl.tensor,
         input_size_h: int, input_size_w: int,
         output_size_h: int, output_size_w: int,
         kernel_size: int, stride: int, padding: int,
         BLOCK_SIZE_KERNEL: tl.constexpr, BLOCK_SIZE_OUTPUT: tl.constexpr,
     ):
         pid = tl.program_id(0)
         slice = tl.program_id(1)
         output_offset = pid * BLOCK_SIZE_OUTPUT + tl.arange(0, BLOCK_SIZE_OUTPUT)
         output_idx_h = output_offset // output_size_w
         output_idx_w = output_offset % output_size_w
         input_idx_h = output_idx_h[:, None] * stride + (tl.arange(0, BLOCK_SIZE_KERNEL) - padding)[None, :]
         input_idx_w = output_idx_w[:, None] * stride + (tl.arange(0, BLOCK_SIZE_KERNEL) - padding)[None, :]
         input_ptrs = input_idx_h[:, :, None] * input_size_w + input_idx_w[:, None, :]
         input_mask_h = (input_idx_h >= 0) & (input_idx_h < input_size_h)
         input_mask_w = (input_idx_w >= 0) & (input_idx_w < input_size_w)
         input_mask = input_mask_h[:, :, None] & input_mask_w[:, None, :]
         kernel_mask = tl.arange(0, BLOCK_SIZE_KERNEL) < kernel_size
         kernel_mask = kernel_mask[:, None] & kernel_mask[None, :]
         input_ptrs = input_ptrs + slice * input_size_h * input_size_w
         input_data = tl.load(input + input_ptrs, mask=input_mask & kernel_mask, other=-float("inf"))
         output_data = tl.max(
             tl.reshape(input_data, (BLOCK_SIZE_OUTPUT, BLOCK_SIZE_KERNEL * BLOCK_SIZE_KERNEL)),
             axis=1, keep_dims=False,
         )
         output_mask = output_offset < output_size_h * output_size_w
         output_ptr = output_offset + slice * output_size_h * output_size_w
         tl.store(output + output_ptr, output_data.to(dtype=output.dtype.element_ty), mask=output_mask)


     # ============ variant: original main path + semantics-no-op injection ==============
     # The injection is the delta-reduced minimal trigger: an opaque (runtime-uniform)
     # prologue, one template block writing only to the scratch buffer, a duplicate
     # output_offset branch, and a dead `return` under the (never-true) opaque_false.
     @triton.jit
     def _helper_arg(_v):
         return tl.argmax(_v, axis=0) + tl.argmin(_v, axis=1)


     @triton.jit
     def _helper_off(pid, BLOCK_SIZE_OUTPUT):
         return pid * BLOCK_SIZE_OUTPUT + tl.arange(0, BLOCK_SIZE_OUTPUT)


     @triton.jit
     def max_pooling_kernel_tx(
         switch_ptr, switch_out_ptr,
         input: tl.tensor, output: tl.tensor,
         input_size_h: int, input_size_w: int,
         output_size_h: int, output_size_w: int,
         kernel_size: int, stride: int, padding: int,
         BLOCK_SIZE_KERNEL: tl.constexpr, BLOCK_SIZE_OUTPUT: tl.constexpr,
     ):
         opaque_values = tl.load(switch_ptr + tl.arange(0, 2))
         (opaque_zero, opaque_one) = tl.split(opaque_values)
         opaque_true = opaque_zero < opaque_one    # 0 < 1 -> always True
         opaque_false = opaque_zero > opaque_one   # always False
         if opaque_true:
             tpl_v = tl.reshape(tl.arange(0, 16), (4, 4)) + opaque_one
             tpl_a = _helper_arg(tpl_v)
             tpl_r = (tl.sum(tpl_a) + opaque_values).to(tl.float32)
             if opaque_true:
                 tl.store(switch_out_ptr + 0 + tl.arange(0, 2), tpl_r,
                          mask=(tl.program_id(axis=0) == 0) & (tl.program_id(axis=1) == 0) & (tl.program_id(axis=2) == 0))
             else:
                 pass
         pid = tl.program_id(0)
         slice = tl.program_id(1)
         if opaque_true:
             output_offset = pid * BLOCK_SIZE_OUTPUT + tl.arange(0, BLOCK_SIZE_OUTPUT)
         else:
             output_offset = _helper_off(pid, BLOCK_SIZE_OUTPUT)
         output_idx_h = output_offset // output_size_w
         output_idx_w = output_offset % output_size_w
         input_idx_h = output_idx_h[:, None] * stride + (tl.arange(0, BLOCK_SIZE_KERNEL) - padding)[None, :]
         input_idx_w = output_idx_w[:, None] * stride + (tl.arange(0, BLOCK_SIZE_KERNEL) - padding)[None, :]
         input_ptrs = input_idx_h[:, :, None] * input_size_w + input_idx_w[:, None, :]
         input_mask_h = (input_idx_h >= 0) & (input_idx_h < input_size_h)
         input_mask_w = (input_idx_w >= 0) & (input_idx_w < input_size_w)
         input_mask = input_mask_h[:, :, None] & input_mask_w[:, None, :]
         kernel_mask = tl.arange(0, BLOCK_SIZE_KERNEL) < kernel_size
         kernel_mask = kernel_mask[:, None] & kernel_mask[None, :]
         input_ptrs = input_ptrs + slice * input_size_h * input_size_w
         if opaque_false:  # dead
             return
         input_data = tl.load(input + input_ptrs, mask=input_mask & kernel_mask, other=-float("inf"))
         output_data = tl.max(
             tl.reshape(input_data, (BLOCK_SIZE_OUTPUT, BLOCK_SIZE_KERNEL * BLOCK_SIZE_KERNEL)),
             axis=1, keep_dims=False,
         )
         output_mask = output_offset < output_size_h * output_size_w
         output_ptr = output_offset + slice * output_size_h * output_size_w
         tl.store(output + output_ptr, output_data.to(dtype=output.dtype.element_ty), mask=output_mask)


     def main() -> int:
         torch.manual_seed(0)
         input_t = torch.randn(100, dtype=torch.float32, device="musa")   # N*C*H*W = 1*1*10*10
         orig = torch.zeros(NUM_OUT, dtype=torch.float32, device="musa")
         trans = torch.zeros(NUM_OUT, dtype=torch.float32, device="musa")
         grid = (triton.cdiv(NUM_OUT, BLOCK_SIZE_OUTPUT), 1)              # (1, 1)
         max_pooling_kernel[grid](input_t, orig, H, W, 5, 5, KS, STRIDE, PADDING,
                                  BLOCK_SIZE_KERNEL=BLOCK_SIZE_KERNEL, BLOCK_SIZE_OUTPUT=BLOCK_SIZE_OUTPUT)
         switch_ptr = torch.tensor([0, 1], dtype=torch.int32, device="musa")
         switch_out = torch.empty(400, dtype=torch.float32, device="musa")
         max_pooling_kernel_tx[grid](switch_ptr, switch_out, input_t, trans, H, W, 5, 5, KS, STRIDE, PADDING,
                                     BLOCK_SIZE_KERNEL=BLOCK_SIZE_KERNEL, BLOCK_SIZE_OUTPUT=BLOCK_SIZE_OUTPUT)
         torch.musa.synchronize()
         o, t = orig.cpu().tolist(), trans.cpu().tolist()
         n_bad = sum(1 for a, b in zip(o, t) if a != b)
         print("original   =", o)
         print("transformed=", t)
         print(f"mismatch = {n_bad}/{NUM_OUT}")
         return 0 if n_bad > 0 else 1


     if __name__ == "__main__":
         raise SystemExit(main())
   ```

 **Details / root-cause analysis** **GERMEATED BY CLAUDE CODE**

   • Source semantics are identical: the template block only writes to the scratch buffer; the main store is line-for-line
     the original kernel. Rebuilding the same template with the original main path (no double-branch, no dead return) yields
     correct output.
   • Not a runtime effect: with switch = [0,0] (template branch compiled but never executed at runtime, main path runs
     through the else branch) the result is still wrong — the defect is in the compiled artifact itself.
   • IR localization: TTIR is byte-identical to the correct A/B build (same source, num_warps=8); TTGIR differs only in
     warpsPerCTA; LLIR differs massively (a 1024-lane masked load scalarized with per-lane -inf phi constants). The
     divergence is in the LLIR / code-generation layer.
   • Heisenbug probe: duplicating the main store to a second ("probe") buffer flips the result back to correct — classic
     register-allocation / scheduling-state sensitivity.
   • Layout sensitivity: 128 elements/warp (BLOCK_SIZE_OUTPUT=512 or num_warps=8) is correct; 256 elements/warp
     (BLOCK_SIZE_OUTPUT=1024, num_warps=4) is wrong.
   • Delta-reduce of the trigger: all five pieces must be present — opaque prologue, template block with ≥1 block
     computation + 1 scratch store, duplicate output_offset branch (either side's expression is fine), dead return (even an
     empty body), and the main path's masked load (other=-inf) + tl.max(axis=1) + store. Removing any one of them fixes the
     output. The template need not be complex (tl.max reduce alone suffices), but its structural form (branch-wrapped, store
     included) matters — a straight-line template does not trigger.

   Workaround

   Not a practical user-facing workaround exists yet; the layout knob that avoids it:

   ```
     BLOCK_SIZE_OUTPUT=512   (128 elements/warp)   -> correct
     num_warps=8             (128 elements/warp)   -> correct for the minimal form
   ```

   Related observations

   This is the same "branch + layout/codegen interaction" family as the nested-scf.if layout-hoist miscompile previously
   reported for a softmax/attention kernel on the same backend, but with a wider trigger surface (no nested if, no
   math.exp). It differs in mechanism from the "opaque-wrapped store skipped" miscompiles (topk / matrix-power variants) and
   from the MTGPU llc compile-time failures (SIGSEGV in PreRA, compile hang, register-allocation error) — here compilation
   succeeds and the executable is wrong.

**Comment by zhzhcookie** [comment]
Please try `flagtree===0.6.2a3+mthreads3.6`, or build flagtree from source.
